implementations/rust/ockam/ockam_api/src/authenticator/direct.rs

```diff
@@ -0,0 +1,419 @@
+                // Admin wants to add an enroller.
+                ["enroller"] if self.is_admin(from) => {
+                    let add: AddEnroller = dec.decode()?;
+                    let enroller = minicbor::to_vec(Placeholder)?;
+                    self.e_store
+                        .set(ENROLLER, add.enroller().as_str().to_string(), enroller)
+                        .await?;
+                    Response::ok(req.id()).to_vec()?
+                }
+                ["enroller"] => Response::unauthorized(req.id()).to_vec()?,
+                // Enroller wants to add a member.
+                ["member"] => {
+                    if let Some(err) = check_enroller(&self.e_store, &req, from.key_id()).await? {
+                        err.to_vec()?
+                    } else {
+                        let add: AddMember = dec.decode()?;
+                        let member = minicbor::to_vec(Placeholder)?;
+                        self.m_store
+                            .set(MEMBER, add.member().as_str().to_string(), member)
+                            .await?;
+                        Response::ok(req.id()).to_vec()?
+                    }
+                }
+                // Member wants a credential.
+                ["credential"] => {
+                    if let Some(err) = check_member(&self.m_store, &req, from.key_id()).await? {
+                        err.to_vec()?
+                    } else {
+                        let mut attrs = Attributes::new();
+                        attrs.put("id", from.key_id().as_bytes());
+                        let ts = Timestamp::now().ok_or_else(invalid_sys_time)?;
+                        let mut timestamp = Cursor::new([0; 9]);
+                        minicbor::encode(ts, &mut timestamp)?;
+                        attrs.put("ts", &timestamp.get_ref()[..timestamp.position()]);
+                        let crd = self.signer.sign(&attrs).await?;
+                        Response::ok(req.id()).body(crd).to_vec()?
+                    }
+                }
+                _ => response::unknown_path(&req).to_vec()?,
+            },
+            Some(Method::Get) => match req.path_segments::<3>().as_slice() {
+                // Admin wants to check enroller data.
+                ["enroller", id] if self.is_admin(from) => {
+                    if let Some(data) = self.e_store.get(ENROLLER, id).await? {
+                        let enroller = minicbor::decode::<Placeholder>(&data)?;
+                        Response::ok(req.id()).body(enroller).to_vec()?
+                    } else {
+                        Response::not_found(req.id()).to_vec()?
+                    }
+                }
+                ["enroller", _] => Response::unauthorized(req.id()).to_vec()?,
+                // Enroller checks member data.
+                ["member", id] => {
+                    if let Some(err) = check_enroller(&self.e_store, &req, from.key_id()).await? {
```

Comment:
Is admin allowed to do everything enroller has permissions to do?

Reply:
Effectively yes. The admin would first have to add itself as an enroller before adding members.